File: Python/Main/Cliente.py
from Transaccion import Transaccion
from TarjetaDebito import TarjetaDebito
from Divisa import Divisa

class Cliente():
    def __init__(self, nombre, noDeIdentificacion):
        from Banco import Banco
        self.nombre = nombre
        self.Id = noDeIdentificacion
        self.tarjetasDebito = []
        self.tarjetasCredito = []
        self.facturas = []
        Banco.agregarCliente(self)
    
    def getNombre(self):
        return self.nombre
    
    def revisarHistorialDeCreditos(self):
        from TarjetaCredito import TarjetaCredito
        retorno = []
        for t in Transaccion.getTransacciones():
            if t.getClienteOrigen() == self and not t.rechazado and isinstance(t.getTarjetaOrigen(), TarjetaCredito) and not t.pendiente:
                retorno.append(t)
        return retorno
    
    def getTarjetasCredito(self):
        return self.tarjetasCredito
    def getTarjetasDebito(self):
        return self.tarjetasDebito
    
    def getFacturas(self):
        return self.facturas
    
    def agregarFactura(self, f):
        self.facturas.append(f)

    def getTarjetas(self):
        return self.tarjetasCredito + self.tarjetasDebito
    
    def listarTarjetas(self, factura):
        tarjetasDisponibles = []
        for tarjeta in self.tarjetasDebito:
            if not tarjeta.tieneSaldo():
                continue
            if tarjeta.getDivisa() == factura.getDIVISA():
                tarjetasDisponibles.append(tarjeta)
        for tarjeta in self.tarjetasCredito:
            if not tarjeta.tieneSaldo():
                continue
            if tarjeta.getDivisa() == factura.getDIVISA():
                tarjetasDisponibles.append(tarjeta)
        return tarjetasDisponibles
    
    def agregarTarjetasDebito(self, *tarjetasDebito):
        self.tarjetasDebito.extend(list(tarjetasDebito))
    
    def agregarTarjetasCredito(self, *tarjetasCredito):
        self.tarjetasCredito.extend(list(tarjetasCredito))

    def encontrarTarjeta(self, tarjeta): #Sirve para encontrar una tarjeta basándose en el toString de esa tarjeta
        for t in self.tarjetasCredito:
            if t.__str__() == tarjeta.__str__():
                return t
        for t in self.tarjetasDebito:
            if t.__str__() == tarjeta.__str__():
                return t
    
    def encontrarFacturas(self, factura):
        for f in self.facturas:
            if f.__str__() == factura.__str__():
                return f

    def mostrarFacturas(self):
        retorno = []
        for f in self.facturas:
            if not f.facturaPagada:
                retorno.append(f)
        return retorno
    
    def verPeticiones(self):
        retorno = []
        for t in Transaccion.getTransacciones():
            if t.getClienteObjetivo() is not None:
                if t.getClienteObjetivo() == self and t.getMensaje() is not None and t.isPendiente():
                    retorno.append(t)
        return retorno
    
    def tarjetasConDivisa(self, divisa, origen: bool): #Devuelve una lista con las tarjetas del cliente que tengan una cierta divisa
        retorno = []
        for t in self.tarjetasDebito:
            if not t.isActiva():
                continue
            if not t.tieneSaldo():
                continue
            if t.getDivisa() == divisa:
                retorno.append(t)
        if origen:
            for t in self.tarjetasCredito:
                if not t.isActiva():
                    continue
                if not t.tieneSaldo():
                    continue
                if t.getDivisa() == divisa:
                    retorno.append(t)
        return retorno
    
    def listarCanales(self, divisao, divisad):
        from Banco import Banco
        divisaOrigen = divisao
        divisaDestino = divisad
        canales = []
        for canal in Banco.getCanales():
            if canal.tieneDivisa(divisaOrigen):
                continue
            if canal.tieneDivisa(divisaDestino):
                continue
            canales.append(canal)
        return Banco.ordenarCanalesPorImpuestos(canales)
    

File: Java/Banco/src/gestorAplicacion/entidades_de_negocio/Cliente.java
//Autores:
//Jose Miguel Pulgarin A.
//Carlos Guarin
//Dario Alexander Penagos V.

package gestorAplicacion.entidades_de_negocio;

import java.io.Serial;
import java.util.Arrays;

import gestorAplicacion.infraestructura.*;
import gestorAplicacion.tarjetas.*;

import java.io.Serializable;
import java.util.ArrayList;

public class Cliente implements Serializable{
	@Serial
	private static final long serialVersionUID = 1L;

	public final String NOMBRE;
	private int Id;
	private ArrayList<TarjetaDebito> tarjetasDebito = new ArrayList<>();
	private ArrayList<TarjetaCredito> tarjetasCredito = new ArrayList<>();
	private ArrayList<Factura> facturas = new ArrayList<>();
	private int bonoActual = 0;
	
	public Cliente(String nombre, int noDeIdentificacion) {
		this.NOMBRE = nombre;
		this.Id = noDeIdentificacion;
		Banco.agregarCliente(this);
	}
	
	
	//Metodos de la instancia
	
	@Override
	public String toString(){
		return "\n\nNombre: %s\nIdentificacion: %s\nNúmero de Tarjetas de Debito: %s\nNúmero de Tarjetas de Credito: %s\nBono Actual: %s".formatted(NOMBRE, Id, tarjetasDebito.size(), tarjetasCredito.size(), bonoActual);
	}

	public ArrayList<TarjetaDebito>  getTarjetasDebito() {
		return tarjetasDebito;
	}

	public void agregarTarjetasDebito(TarjetaDebito... tarjetasDebito) {//Agregar varios tarjetas una por una Ej. (tarjeta1, tarjeta2, etc)
		this.tarjetasDebito.addAll(Arrays.asList(tarjetasDebito));
	}

	public ArrayList<TarjetaCredito> getTarjetasCredito() {
		return tarjetasCredito;
	}

	public ArrayList<Tarjeta> getTarjetas() {
		ArrayList<Tarjeta> retorno = new ArrayList<>();
		retorno.addAll(tarjetasDebito);
		retorno.addAll(tarjetasCredito);
		return retorno;
	}

	public void agregarTarjetasCredito(TarjetaCredito... tarjetasCredito) {//Agregar varios tarjetas una por una Ej. (tarjeta1, tarjeta2, etc)
		this.tarjetasCredito.addAll(Arrays.asList(tarjetasCredito));
	}

	public ArrayList<Tarjeta> seleccionarTarjeta(Divisa divisa, boolean retirar){
		ArrayList<Tarjeta> retorno = new ArrayList<>();
		for(TarjetaDebito t : tarjetasDebito){
			if(t.getDivisa().equals(divisa)) retorno.add(t);
		}
		if(!retirar){
			for(TarjetaCredito t : tarjetasCredito){
				if(t.getDivisa().equals(divisa)) retorno.add(t);
			}
		}
		return retorno;
	}

	public ArrayList<Factura> getFactura() {
		return facturas;
	}

	public void agregarFactura(Factura factura) {
		this.facturas.add(factura);
	}

	public int getBonoActual() {
		return bonoActual;
	}

	public void setBonoActual(int bonoActual) {
		this.bonoActual = bonoActual;
	}
	
	//Metodos de las intancias
	
	public ArrayList<Factura> listarFacturas(){//Este metodo solo retorna facturas pendientes
		ArrayList<Factura> retorno = new ArrayList<>();
		for(Factura f : this.facturas){
			if(f.isFacturaVencida() && f.isFacturaPagada()){
				retorno.add(f);
			}
		}
		for(Factura f : this.facturas){
			if(!f.isFacturaVencida() && f.isFacturaPagada()){
				retorno.add(f);
			}
		}
		return retorno;
	}
	
	//Retorna las tarjetas de credito y debito que sean compatibles con la divisa de la factura
	//Ademas que no tengan fondos en 0 o creditoMaximo alcanzado
	//La tarjeta para ser listada también debe tener un estado "Activo"
	public ArrayList<Tarjeta> listarTarjetas(Factura factura){//Listar tarjetas para pagar factura
		ArrayList<Tarjeta> tarjetasDisponibles = new ArrayList<>();
		for(TarjetaDebito tarjeta : tarjetasDebito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(factura.getDIVISA())){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		for(TarjetaCredito tarjeta : tarjetasCredito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(factura.getDIVISA())){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		return tarjetasDisponibles;
	}
	
	public ArrayList<Tarjeta> listarTarjetas(ArrayList<Divisa> divisas){//Listar tarjetas para cambiar divisas. Recibe un arrayList
		Divisa divisaOrigen = divisas.get(0);
		Divisa divisaDestino = divisas.get(1);
		
		ArrayList<Tarjeta> tarjetasDisponibles = new ArrayList<>();
		
		for(TarjetaDebito tarjeta : tarjetasDebito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(divisaOrigen) || tarjeta.getDivisa().equals(divisaDestino)){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		
		for(TarjetaCredito tarjeta : tarjetasCredito){
			if(!tarjeta.isActiva())
				continue;
			if(!tarjeta.tieneSaldo())
				continue;
			if(tarjeta.getDivisa().equals(divisaOrigen) || tarjeta.getDivisa().equals(divisaDestino)){
				tarjetasDisponibles.add(tarjeta);
			}
		}
		
		return tarjetasDisponibles;
	}
	
	//El usuario debe tener tarjetas con la divisa de origen que escoja
	//Dichas tarjetas deben tener saldo o credito, y estar activas
	public ArrayList<Divisa> escogerDivisas(Divisa origen, Divisa destino) {//Retorna un ArrayList
		ArrayList<Divisa> divisas = new ArrayList<>();
		divisas.add(origen);
		divisas.add(destino);
		
		//Verificar que el cliente tenga una tarjeta de origen con la divisa que escogió
		for(TarjetaDebito tarjetaDebito: this.getTarjetasDebito()) {
			if(!tarjetaDebito.isActiva())
				continue;
			if(!tarjetaDebito.tieneSaldo())
				continue;
			if(!tarjetaDebito.getDivisa().equals(origen))
				continue;
			//Si existe al menos una tarjeta que cumpla con lo requerido, retornamos.
			return divisas;
		}
		
		//Si no existen aun tarjetas con dicha divisa (ya que no se retornó) se recorren las tarjetas de credito
		for(TarjetaCredito tarjetaCredito: this.getTarjetasCredito()) {
			if(!tarjetaCredito.isActiva())
				continue;
			if(!tarjetaCredito.tieneSaldo())
				continue;
			if(!tarjetaCredito.getDivisa().equals(origen))
				continue;
			return divisas;
		}
		
		//Si no se retornó anteriormente, es porque el cliente no tiene tarjetas con dicha divisa
		return null;
	}

	public ArrayList<Transaccion> verPeticiones(){
		ArrayList<Transaccion> retorno = new ArrayList<>();
		for(Transaccion t : Transaccion.getTransacciones()){
			if(t.getClienteObjetivo() != null){		//Verifica si la transacción tiene un cliente objetivo. sin esta verificación, la siguiente línea podría soltar un error
				if(t.getClienteObjetivo().equals(this) && t.getMensaje() != null && t.isPendiente()){
					retorno.add(t);
				}
			}
			
		}
		return retorno;
	}

	public ArrayList<Canal> listarCanales(ArrayList<Divisa> divisas) {//recibe un arrayList
		Divisa divisaOrigen = divisas.get(0);
		Divisa divisaDestino = divisas.get(1);
		
		ArrayList<Canal> canales = new ArrayList<>();
		for(Canal canal: Banco.getCanales()) {
			if(canal.tieneDivisa(divisaOrigen))
				continue;
			if(canal.tieneDivisa(divisaDestino))
				continue;
			if(canal.tieneFondosDeDivisa(divisaDestino))
				continue;
			canales.add(canal);
		}
		
		//ordenar por menos impuestos de menor a mayor
		return Banco.ordenarCanalesPorImpuestos(canales);
	}
	
	public ArrayList<Transaccion> revisarHistorialCreditos(){
		ArrayList<Transaccion> retorno = new ArrayList<>();
		for(Transaccion t : Transaccion.getTransacciones()){
			if(t.getClienteOrigen().equals(this) && t.getTarjetaOrigen() instanceof TarjetaCredito && !t.isRechazado() && !t.isPendiente()){
				retorno.add(t);
			}
		}
		return retorno;
	}
}
